README.txt
# Milestone Project 1 - Building a Platform scroller game
by Robert Behm

### Description
This is a platform scroller game where you try to navigate the character across the terrain without falling or dying. 
Navigation is done using the left and right arrow keys, and jumping is done with the up arrow key. 

### Demo
There is not a working demo for this game yet.

### Technologies
This project was created in javascript with HTML Canvas.

### Technical Information
The current implementation is run using VS Code's "Live Server"

### Issues

At the time of presentation, I still have a few things to build out:
NPC characters
Mystery Boxes w/ power ups
moving platforms
attack function for main character
and improved win animation
multiple levels
music and sound effects
health

### Prologue:
As a kid born in the early ninties, I discovered "Super Mario Land: 6 Golden Coins" 
on Game Boy and it quickly became my favorite game for a decade. A few decades later, 
I was very excited to get a chance to learn how to creaste a platform game of my own.

I first brainstormed the key components of the game and I came up with the following:
Character - move, jump, duck, attack
Terrain - floor, platforms, background, blocks, fire/spikes/death pits
NPC Charaters - move back and forth, make main character lose health
Events - die, level up, lose health, kill npc, beat level
Create Parallax scroll effect

I started the game development and attempted to use a move function code that manipulated the position of the player on the screen. I quickly found out that just moving position would not work because the character would not be able to jump and move right or left at the same time. I need to create a velocity component and add gravity to make a playable experience. 

I was also exploring how to create the illusion that the character was moving through the level. I found out that a paralax scroll effect in combination with moving the terrain was the way to make this happen. 

At this point I found this tutorial on how to build a platform game:
https://www.youtube.com/watch?v=4q2vvZn5aoo
In this video, Chris walks through the logic and process to build a platform game. This video was helpful for me to learn how to use Canvas, walk though how to build the physics, collision detection, parallax scroll, and sprite character animation. 

In the tutorial, Chris uses a separate library to crate a live server to display his images, but I was able to use VS Code's 'Live Server" feature to do the same thing. 

I was able to find copyright free graphics for the parallax background and the sprite animated character from Itch.io. 

The parallax images I found had 5 different layers included, which create a sprauling hillside illusion when each of the distant layers scroll at a slower speed than the near layers. 

The sprite character was also very interesting to learn about because the attached image is a series of still images of the character that make the character come to life when cycled through quickly. 

I really enjoyed this project because it challenged me to learn new things to create my game while deepening my confidence with javascript and thinking like a programmer.



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [Unreleased]

## [1.0] - 5/19/2022
### Project submitted for Milestone Project 1
-Character can move back and forth, and jump
-Character can fall into the death pits
-Game restarts when character falls
-Game notifys the player when they have competed the level
-platforms are built out to jump in between
-paralax background creates a depth effect
-Character is animated with sprite animations

backup.js
function newImage(url, width, height) {
  let image = document.createElement("img");
  image.src = url;
  image.width = width;
  image.height = height;
  image.style.position = "absolute";
  document.body.append(image);
  return image;
}

const dirt1 = newImage("./assets/dirt1.png", 0, 0);
const dirt2 = newImage("./assets/dirt2.png", 0, 0);
const dirt3 = newImage("./assets/dirt3.png", 0, 0);
const grass1 = newImage("./assets/grass1.png", 0, 0);
const grass2 = newImage("./assets/grass2.png", 0, 0);
const grass3 = newImage("./assets/grass3.png", 0, 0);
const hills11 = newImage("./assets/hills1-1.png", 0, 0);
const hills12 = newImage("./assets/hills1-2.png", 0, 0);
const hills13 = newImage("./assets/hills1-3.png", 0, 0);
const hills21 = newImage("./assets/hills2-1.png", 0, 0);
const hills22 = newImage("./assets/hills2-2.png", 0, 0);
const hills23 = newImage("./assets/hills2-3.png", 0, 0);
const hills31 = newImage("./assets/hills3-1.png", 0, 0);
const hills32 = newImage("./assets/hills3-2.png", 0, 0);
const hills33 = newImage("./assets/hills3-3.png", 0, 0);
const sky1 = newImage("./assets/sky1.png", 0, 0);
const sky2 = newImage("./assets/sky2.png", 0, 0);
const sky3 = newImage("./assets/sky3.png", 0, 0);

const idleRight = newImage("./assets/IdleRight.png", 0, 0);
const idleLeft = newImage("./assets/IdleLeft.png", 0, 0);
const runRight = newImage("./assets/RunRight.png", 0, 0);
const runLeft = newImage("./assets/RunLeft.png", 0, 0);

const youWin = newImage("./assets/youwin.jpg", 0, 0);

const canvas = document.querySelector("canvas");
const ctx = canvas.getContext("2d");

canvas.width = window.innerWidth;
canvas.height = 600;

const gravity = 1.5;

class Player {
  constructor() {
    this.speed = 6;
    this.position = {
      x: window.innerWidth * 0.25,
      y: 0,
    };
    this.velocity = {
      x: 0,
      y: 0,
    };
    this.width = 120;
    this.height = 120;
    this.image = idleRight;
    this.frames = 0;
    this.sprites = {
      idle: {
        right: idleRight,
        left: idleLeft,
        cropWidth: 510,
      },
      run: {
        right: runRight,
        left: runLeft,
      },
    };
    this.currentSprite = this.sprites.idle.right;
    this.currentCropWidth = 510;
  }

  draw() {
    ctx.drawImage(
      this.currentSprite,
      this.currentCropWidth * this.frames,
      0,
      340,
      400,
      this.position.x,
      this.position.y,
      this.width,
      this.height
    );
  }
  update() {
    this.frames++;
    //console.log(this.frames);
    if (
      this.frames >= 16 &&
      (this.currentSprite === this.sprites.idle.right ||
        this.currentSprite === this.sprites.idle.left)
    )
      this.frames = 0;
    else if (
      this.frames >= 13 &&
      (this.currentSprite === this.sprites.run.right ||
        this.currentSprite === this.sprites.run.left)
    )
      this.frames = 0;
    this.draw();
    this.position.y += this.velocity.y;
    this.position.x += this.velocity.x;

    if (
      this.position.y + this.height + this.velocity.y <=
      canvas.height + player.height
    ) {
      this.velocity.y += gravity;
    }
    // else {this.velocity.y = 0;  }
  }
}
class Platform {
  constructor({ image, x, y, width, height }) {
    this.position = {
      x: x,
      y: y,
    };
    this.image = image;
    this.width = width;
    this.height = height;
  }

  draw() {
    // ctx.fillStyle = "brown";
    // ctx.fillRect(this.position.x, this.position.y, this.width, this.height);
    ctx.drawImage(
      this.image,
      this.position.x,
      this.position.y,
      this.width,
      this.height
    );
  }
}
class Background {
  constructor({ image, x, y, width, height, scrollSpeed }) {
    this.position = {
      x: x,
      y: y,
    };
    this.image = image;
    this.width = width;
    this.height = height;
  }

  draw() {
    // ctx.fillStyle = "brown";
    // ctx.fillRect(this.position.x, this.position.y, this.width, this.height);
    ctx.drawImage(
      this.image,
      this.position.x,
      this.position.y,
      this.width,
      this.height
    );
  }
}

let player = new Player();
// console.log(player.sprites.run.right);
let platforms = [];

let skyBackground = [];
let hill3Background = [];
let hill2Background = [];
let hill1Background = [];
let grassBackground = [];
let youWinPhoto;
let scrollOffset = 0;

function start() {
  player = new Player();

  platforms = [
    new Platform({ image: dirt1, x: 0, y: 520, width: 800, height: 80 }),
    new Platform({ image: dirt2, x: 799, y: 520, width: 800, height: 80 }),
    new Platform({ image: dirt3, x: 1950, y: 520, width: 800, height: 80 }),
    new Platform({ image: dirt3, x: 1670, y: 400, width: 200, height: 20 }),
    new Platform({ image: dirt3, x: 2670, y: 400, width: 200, height: 20 }),
    new Platform({ image: dirt3, x: 2870, y: 270, width: 200, height: 20 }),
    new Platform({ image: dirt3, x: 3200, y: 270, width: 40, height: 20 }),
    new Platform({ image: dirt3, x: 3400, y: 270, width: 40, height: 20 }),
    new Platform({ image: dirt3, x: 3500, y: 400, width: 100, height: 20 }),
    new Platform({ image: dirt3, x: 3600, y: 520, width: 800, height: 80 }),
    new Platform({ image: dirt3, x: 4400 - 2, y: 520, width: 800, height: 80 }),
    new Platform({ image: dirt3, x: 5400, y: 520, width: 40, height: 20 }),
    new Platform({ image: dirt3, x: 5400, y: 400, width: 40, height: 20 }),
    new Platform({ image: dirt3, x: 5550, y: 270, width: 400, height: 20 }),
    new Platform({ image: dirt3, x: 6100, y: 150, width: 400, height: 20 }),
    new Platform({ image: dirt3, x: 6700, y: 520, width: 800, height: 80 }),
    new Platform({ image: dirt3, x: 7500 - 2, y: 520, width: 800, height: 80 }),
  ];

  skyBackground = [
    new Background({ image: sky1, x: 0, y: 0, width: 1200, height: 600 }),
    new Background({ image: sky2, x: 1198, y: 0, width: 1200, height: 600 }),
    new Background({ image: sky3, x: 2396, y: 0, width: 1200, height: 600 }),
  ];
  hill3Background = [
    new Background({ image: hills31, x: 0, y: 180, width: 1200, height: 400 }),
    new Background({
      image: hills32,
      x: 1199,
      y: 180,
      width: 1200,
      height: 400,
    }),
    new Background({
      image: hills33,
      x: 2398,
      y: 180,
      width: 1200,
      height: 400,
    }),
    new Background({
      image: hills31,
      x: 3597,
      y: 180,
      width: 1200,
      height: 400,
    }),
    new Background({
      image: hills32,
      x: 4796,
      y: 180,
      width: 1200,
      height: 400,
    }),
    new Background({
      image: hills33,
      x: 5995,
      y: 180,
      width: 1200,
      height: 400,
    }),
  ];
  hill2Background = [
    new Background({ image: hills21, x: 0, y: 220, width: 1200, height: 400 }),
    new Background({
      image: hills22,
      x: 1199,
      y: 220,
      width: 1200,
      height: 400,
    }),
    new Background({
      image: hills23,
      x: 2398,
      y: 205,
      width: 1200,
      height: 400,
    }),
    new Background({
      image: hills21,
      x: 3597,
      y: 220,
      width: 1200,
      height: 400,
    }),
    new Background({
      image: hills22,
      x: 4796,
      y: 220,
      width: 1200,
      height: 400,
    }),
    new Background({
      image: hills23,
      x: 5995,
      y: 205,
      width: 1200,
      height: 400,
    }),
  ];
  hill1Background = [
    new Background({ image: hills11, x: 0, y: 240, width: 1200, height: 400 }),
    new Background({
      image: hills12,
      x: 1199,
      y: 276,
      width: 1200,
      height: 400,
    }),
    new Background({
      image: hills13,
      x: 2398,
      y: 286,
      width: 1200,
      height: 400,
    }),
    new Background({
      image: hills11,
      x: 3597,
      y: 240,
      width: 1200,
      height: 400,
    }),
    new Background({
      image: hills12,
      x: 4797,
      y: 276,
      width: 1200,
      height: 400,
    }),
    new Background({
      image: hills13,
      x: 5996,
      y: 286,
      width: 1200,
      height: 400,
    }),
  ];
  grassBackground = [
    new Background({ image: grass1, x: 0, y: 200, width: 900, height: 400 }),
    new Background({ image: grass2, x: 898, y: 466, width: 900, height: 150 }),
    new Background({ image: grass3, x: 1797, y: 307, width: 900, height: 300 }),
    new Background({ image: grass1, x: 2696, y: 200, width: 900, height: 400 }),
    new Background({ image: grass2, x: 3595, y: 466, width: 900, height: 150 }),
    new Background({ image: grass3, x: 4494, y: 307, width: 900, height: 300 }),
    new Background({ image: grass1, x: 5393, y: 200, width: 900, height: 400 }),
    new Background({ image: grass2, x: 6292, y: 466, width: 900, height: 150 }),
    new Background({ image: grass3, x: 7190, y: 307, width: 900, height: 300 }),
  ];
  youWinPhoto = new Background({
    image: youWin,
    x: 150,
    y: 50,
    width: 852,
    height: 480,
  });
  scrollOffset = 0;
}
let lastKey;
const keys = {
  right: {
    pressed: false,
  },
  left: {
    pressed: false,
  },
};

function animate() {
  requestAnimationFrame(animate);
  ctx.fillStyle = "white";
  ctx.clearRect(0, 0, canvas.width, canvas.height);
  skyBackground.forEach((sky) => {
    sky.draw();
  });
  hill3Background.forEach((hill) => {
    hill.draw();
  });
  hill2Background.forEach((hill) => {
    hill.draw();
  });
  hill1Background.forEach((hill) => {
    hill.draw();
  });
  grassBackground.forEach((grass) => {
    grass.draw();
  });
  platforms.forEach((platform) => {
    platform.draw();
  });

  //   console.log(player.position.x + scrollOffset);
  player.update();
  if (
    keys.right.pressed == true &&
    player.position.x < window.innerWidth * (2 / 3)
  ) {
    player.velocity.x = player.speed;
  } else if (
    (keys.left.pressed == true &&
      player.position.x >= window.innerWidth * (1 / 3)) ||
    (keys.left.pressed && scrollOffset === 0 && player.position.x > 0)
  ) {
    player.velocity.x = -player.speed;
  } else {
    player.velocity.x = 0;

    if (keys.right.pressed) {
      scrollOffset += player.speed;
      skyBackground.forEach((sky) => {
        sky.position.x -= player.speed * 0.2;
      });
      hill3Background.forEach((hill) => {
        hill.position.x -= player.speed * 0.4;
      });
      hill2Background.forEach((hill) => {
        hill.position.x -= player.speed * 0.5;
      });
      hill1Background.forEach((hill) => {
        hill.position.x -= player.speed * 0.6;
      });
      grassBackground.forEach((grass) => {
        grass.position.x -= player.speed * 0.7;
      });
      platforms.forEach((platform) => {
        platform.position.x -= player.speed;
      });
    } else if (keys.left.pressed && scrollOffset > 0) {
      scrollOffset -= player.speed;
      skyBackground.forEach((sky) => {
        sky.position.x += player.speed * 0.2;
      });
      hill3Background.forEach((hill) => {
        hill.position.x += player.speed * 0.4;
      });
      hill2Background.forEach((hill) => {
        hill.position.x += player.speed * 0.5;
      });
      hill1Background.forEach((hill) => {
        hill.position.x += player.speed * 0.6;
      });
      grassBackground.forEach((grass) => {
        grass.position.x += player.speed * 0.7;
      });
      platforms.forEach((platform) => {
        platform.position.x += player.speed;
      });
    }
  }
  //this is the platform collision detection code
  platforms.forEach((platform) => {
    if (
      player.position.y + player.height <= platform.position.y &&
      player.position.y + player.height + player.velocity.y >=
        platform.position.y &&
      player.position.x + player.width >= platform.position.x &&
      player.position.x <= platform.position.x + platform.width
    ) {
      player.velocity.y = 0;
    }
  });

  if (player.position.y < 0) {
    player.velocity.y = 10;
  }

  if (
    keys.right.pressed &&
    lastKey === "right" &&
    player.currentSprite !== player.sprites.run.right
  ) {
    player.frames = 1;
    player.currentSprite = player.sprites.run.right;
  } else if (
    keys.left.pressed &&
    lastKey === "left" &&
    player.currentSprite !== player.sprites.run.left
  ) {
    player.frames = 1;
    player.currentSprite = player.sprites.run.left;
  } else if (
    !keys.left.pressed &&
    lastKey === "left" &&
    player.currentSprite !== player.sprites.idle.left
  ) {
    player.frames = 1;
    player.currentSprite = player.sprites.idle.left;
  } else if (
    !keys.right.pressed &&
    lastKey === "right" &&
    player.currentSprite !== player.sprites.idle.right
  ) {
    player.frames = 1;
    player.currentSprite = player.sprites.idle.right;
  }

  //win condition
  if (scrollOffset + canvas.width > 8300) {
    //console.log("You Win!");
    youWinPhoto.draw();
    // ctx.font = "48px serif"
    // ctx.fillText("You Win!", 8300, 200, 400 )
  }
  //lose condition
  if (player.position.y > canvas.height) {
    console.log("you lose");
    start();
  }
}
start();
animate();

document.addEventListener("keydown", ({ keyCode }) => {
  //   console.log(event);
  switch (keyCode) {
    case 65:
      keys.left.pressed = true;
      lastKey = "left";
      break;
    case 37:
      keys.left.pressed = true;
      lastKey = "left";
      break;

    case 83:
      //   console.log("down");
      break;
    case 40:
      //   console.log("down");
      break;

    case 68:
      keys.right.pressed = true;
      lastKey = "right";
      break;
    case 39:
      keys.right.pressed = true;
      lastKey = "right";
      break;

    case 87:
      player.velocity.y -= 20;
      break;
    case 38:
      player.velocity.y -= 20;
      break;
  }
});

document.addEventListener("keyup", ({ keyCode }) => {
  //   console.log(event);
  switch (keyCode) {
    case 65:
      keys.left.pressed = false;
      break;
    case 37:
      keys.left.pressed = false;
      break;

    case 83:
      player.velocity.y = 0;
      break;
    case 40:
      player.velocity.y = 0;
      break;

    case 68:
      keys.right.pressed = false;
      break;
    case 39:
      keys.right.pressed = false;
      break;

    case 87:
      player.velocity.y = 0;
      break;
    case 38:
      player.velocity.y = 0;
      break;
  }
});


script.js
//============================Create Images====================================
function newImage(url, width, height) {
  let image = document.createElement("img");
  image.src = url;
  image.width = width;
  image.height = height;
  image.style.position = "absolute";
  document.body.append(image);
  return image;
}
//Paralax Background Images
const dirt1 = newImage("./assets/dirt1.png", 0, 0);
const dirt2 = newImage("./assets/dirt2.png", 0, 0);
const dirt3 = newImage("./assets/dirt3.png", 0, 0);
const grass1 = newImage("./assets/grass1.png", 0, 0);
const grass2 = newImage("./assets/grass2.png", 0, 0);
const grass3 = newImage("./assets/grass3.png", 0, 0);
const hills11 = newImage("./assets/hills1-1.png", 0, 0);
const hills12 = newImage("./assets/hills1-2.png", 0, 0);
const hills13 = newImage("./assets/hills1-3.png", 0, 0);
const hills21 = newImage("./assets/hills2-1.png", 0, 0);
const hills22 = newImage("./assets/hills2-2.png", 0, 0);
const hills23 = newImage("./assets/hills2-3.png", 0, 0);
const hills31 = newImage("./assets/hills3-1.png", 0, 0);
const hills32 = newImage("./assets/hills3-2.png", 0, 0);
const hills33 = newImage("./assets/hills3-3.png", 0, 0);
const sky1 = newImage("./assets/sky1.png", 0, 0);
const sky2 = newImage("./assets/sky2.png", 0, 0);
const sky3 = newImage("./assets/sky3.png", 0, 0);

//Sprite Character Images
const idleRight = newImage("./assets/IdleRight.png", 0, 0);
const idleLeft = newImage("./assets/IdleLeft.png", 0, 0);
const runRight = newImage("./assets/RunRight.png", 0, 0);
const runLeft = newImage("./assets/RunLeft.png", 0, 0);

//Win Condition Image
const youWin = newImage("./assets/youwin.jpg", 0, 0);

// Define canvas Parameters
const canvas = document.querySelector("canvas");
const ctx = canvas.getContext("2d");
canvas.width = window.innerWidth;
canvas.height = 600;
const gravity = 1.5;

//====================Define classes for player, platforms and backgrounds================
//Define Player
class Player {
  constructor() {
    this.speed = 6;
    this.position = {
      x: window.innerWidth * 0.25,
      y: 0,
    };
    this.velocity = {
      x: 0,
      y: 0,
    };
    this.width = 120;
    this.height = 120;
    this.image = idleRight;
    this.frames = 0;
    this.sprites = {
      idle: {
        right: idleRight,
        left: idleLeft,
        cropWidth: 510,
      },
      run: {
        right: runRight,
        left: runLeft,
      },
    };
    this.currentSprite = this.sprites.idle.right;
    this.currentCropWidth = 510;
  }
  draw() {
    ctx.drawImage(
      this.currentSprite,
      this.currentCropWidth * this.frames,
      0,
      340,
      400,
      this.position.x,
      this.position.y,
      this.width,
      this.height
    );
  }
  //
  update() {
    this.frames++;
    if (
      this.frames >= 16 &&
      (this.currentSprite === this.sprites.idle.right ||
        this.currentSprite === this.sprites.idle.left)
    )
      this.frames = 0;
    else if (
      this.frames >= 13 &&
      (this.currentSprite === this.sprites.run.right ||
        this.currentSprite === this.sprites.run.left)
    )
      this.frames = 0;
    this.draw();
    this.position.y += this.velocity.y;
    this.position.x += this.velocity.x;

    if (
      this.position.y + this.height + this.velocity.y <=
      canvas.height + player.height
    ) {
      this.velocity.y += gravity;
    }
    // else {this.velocity.y = 0;  }
  }
}
class Platform {
  constructor({ image, x, y, width, height }) {
    this.position = {
      x: x,
      y: y,
    };
    this.image = image;
    this.width = width;
    this.height = height;
  }

  draw() {
    // ctx.fillStyle = "brown";
    // ctx.fillRect(this.position.x, this.position.y, this.width, this.height);
    ctx.drawImage(
      this.image,
      this.position.x,
      this.position.y,
      this.width,
      this.height
    );
  }
}
class Background {
  constructor({ image, x, y, width, height, scrollSpeed }) {
    this.position = {
      x: x,
      y: y,
    };
    this.image = image;
    this.width = width;
    this.height = height;
  }

  draw() {
    // ctx.fillStyle = "brown";
    // ctx.fillRect(this.position.x, this.position.y, this.width, this.height);
    ctx.drawImage(
      this.image,
      this.position.x,
      this.position.y,
      this.width,
      this.height
    );
  }
}
//initially define all game elements
let player = new Player();
let platforms = [];
let skyBackground = [];
let hill3Background = [];
let hill2Background = [];
let hill1Background = [];
let grassBackground = [];
let youWinPhoto;
let scrollOffset = 0;

//using a "start" function so I can re-run the initalize code if the character dies (and goes back to the beginning)
function start() {
  player = new Player();
  //Paralax background has 5 components : sky, hill3, hill2, hill1, grass and dirt
  platforms = [
    new Platform({ image: dirt1, x: 0, y: 520, width: 800, height: 80 }),
    new Platform({ image: dirt2, x: 799, y: 520, width: 800, height: 80 }),
    new Platform({ image: dirt3, x: 1950, y: 520, width: 800, height: 80 }),
    new Platform({ image: dirt3, x: 1670, y: 400, width: 200, height: 20 }),
    new Platform({ image: dirt3, x: 2670, y: 400, width: 200, height: 20 }),
    new Platform({ image: dirt3, x: 2870, y: 270, width: 200, height: 20 }),
    new Platform({ image: dirt3, x: 3200, y: 270, width: 40, height: 20 }),
    new Platform({ image: dirt3, x: 3400, y: 270, width: 40, height: 20 }),
    new Platform({ image: dirt3, x: 3500, y: 400, width: 100, height: 20 }),
    new Platform({ image: dirt3, x: 3600, y: 520, width: 800, height: 80 }),
    new Platform({ image: dirt3, x: 4400 - 2, y: 520, width: 800, height: 80 }),
    new Platform({ image: dirt3, x: 5400, y: 520, width: 40, height: 20 }),
    new Platform({ image: dirt3, x: 5400, y: 400, width: 40, height: 20 }),
    new Platform({ image: dirt3, x: 5550, y: 270, width: 400, height: 20 }),
    new Platform({ image: dirt3, x: 6100, y: 150, width: 400, height: 20 }),
    new Platform({ image: dirt3, x: 6700, y: 520, width: 800, height: 80 }),
    new Platform({ image: dirt3, x: 7500 - 2, y: 520, width: 800, height: 80 }),
  ];

  skyBackground = [
    new Background({ image: sky1, x: 0, y: 0, width: 1200, height: 600 }),
    new Background({ image: sky2, x: 1198, y: 0, width: 1200, height: 600 }),
    new Background({ image: sky3, x: 2396, y: 0, width: 1200, height: 600 }),
  ];
  hill3Background = [
    new Background({ image: hills31, x: 0, y: 180, width: 1200, height: 400 }),
    new Background({
      image: hills32,
      x: 1199,
      y: 180,
      width: 1200,
      height: 400,
    }),
    new Background({
      image: hills33,
      x: 2398,
      y: 180,
      width: 1200,
      height: 400,
    }),
    new Background({
      image: hills31,
      x: 3597,
      y: 180,
      width: 1200,
      height: 400,
    }),
    new Background({
      image: hills32,
      x: 4796,
      y: 180,
      width: 1200,
      height: 400,
    }),
    new Background({
      image: hills33,
      x: 5995,
      y: 180,
      width: 1200,
      height: 400,
    }),
  ];
  hill2Background = [
    new Background({ image: hills21, x: 0, y: 220, width: 1200, height: 400 }),
    new Background({
      image: hills22,
      x: 1199,
      y: 220,
      width: 1200,
      height: 400,
    }),
    new Background({
      image: hills23,
      x: 2398,
      y: 205,
      width: 1200,
      height: 400,
    }),
    new Background({
      image: hills21,
      x: 3597,
      y: 220,
      width: 1200,
      height: 400,
    }),
    new Background({
      image: hills22,
      x: 4796,
      y: 220,
      width: 1200,
      height: 400,
    }),
    new Background({
      image: hills23,
      x: 5995,
      y: 205,
      width: 1200,
      height: 400,
    }),
  ];
  hill1Background = [
    new Background({ image: hills11, x: 0, y: 240, width: 1200, height: 400 }),
    new Background({
      image: hills12,
      x: 1199,
      y: 276,
      width: 1200,
      height: 400,
    }),
    new Background({
      image: hills13,
      x: 2398,
      y: 286,
      width: 1200,
      height: 400,
    }),
    new Background({
      image: hills11,
      x: 3597,
      y: 240,
      width: 1200,
      height: 400,
    }),
    new Background({
      image: hills12,
      x: 4797,
      y: 276,
      width: 1200,
      height: 400,
    }),
    new Background({
      image: hills13,
      x: 5996,
      y: 286,
      width: 1200,
      height: 400,
    }),
  ];
  grassBackground = [
    new Background({ image: grass1, x: 0, y: 200, width: 900, height: 400 }),
    new Background({ image: grass2, x: 898, y: 466, width: 900, height: 150 }),
    new Background({ image: grass3, x: 1797, y: 307, width: 900, height: 300 }),
    new Background({ image: grass1, x: 2696, y: 200, width: 900, height: 400 }),
    new Background({ image: grass2, x: 3595, y: 466, width: 900, height: 150 }),
    new Background({ image: grass3, x: 4494, y: 307, width: 900, height: 300 }),
    new Background({ image: grass1, x: 5393, y: 200, width: 900, height: 400 }),
    new Background({ image: grass2, x: 6292, y: 466, width: 900, height: 150 }),
    new Background({ image: grass3, x: 7190, y: 307, width: 900, height: 300 }),
  ];
  youWinPhoto = new Background({
    image: youWin,
    x: 150,
    y: 50,
    width: 852,
    height: 480,
  });
  scrollOffset = 0;
}
//keys function keeps track of what key is being pressed
let lastKey;
const keys = {
  right: {
    pressed: false,
  },
  left: {
    pressed: false,
  },
};
//animate function
function animate() {
  requestAnimationFrame(animate);
  ctx.fillStyle = "white";
  ctx.clearRect(0, 0, canvas.width, canvas.height);
  skyBackground.forEach((sky) => {
    sky.draw();
  });
  hill3Background.forEach((hill) => {
    hill.draw();
  });
  hill2Background.forEach((hill) => {
    hill.draw();
  });
  hill1Background.forEach((hill) => {
    hill.draw();
  });
  grassBackground.forEach((grass) => {
    grass.draw();
  });
  platforms.forEach((platform) => {
    platform.draw();
  });

  //   console.log(player.position.x + scrollOffset);
  player.update();
  if (
    keys.right.pressed == true &&
    player.position.x < window.innerWidth * (2 / 3)
  ) {
    player.velocity.x = player.speed;
  } else if (
    (keys.left.pressed == true &&
      player.position.x >= window.innerWidth * (1 / 3)) ||
    (keys.left.pressed && scrollOffset === 0 && player.position.x > 0)
  ) {
    player.velocity.x = -player.speed;
  } else {
    player.velocity.x = 0;

    if (keys.right.pressed) {
      scrollOffset += player.speed;
      skyBackground.forEach((sky) => {
        sky.position.x -= player.speed * 0.2;
      });
      hill3Background.forEach((hill) => {
        hill.position.x -= player.speed * 0.4;
      });
      hill2Background.forEach((hill) => {
        hill.position.x -= player.speed * 0.5;
      });
      hill1Background.forEach((hill) => {
        hill.position.x -= player.speed * 0.6;
      });
      grassBackground.forEach((grass) => {
        grass.position.x -= player.speed * 0.7;
      });
      platforms.forEach((platform) => {
        platform.position.x -= player.speed;
      });
    } else if (keys.left.pressed && scrollOffset > 0) {
      scrollOffset -= player.speed;
      skyBackground.forEach((sky) => {
        sky.position.x += player.speed * 0.2;
      });
      hill3Background.forEach((hill) => {
        hill.position.x += player.speed * 0.4;
      });
      hill2Background.forEach((hill) => {
        hill.position.x += player.speed * 0.5;
      });
      hill1Background.forEach((hill) => {
        hill.position.x += player.speed * 0.6;
      });
      grassBackground.forEach((grass) => {
        grass.position.x += player.speed * 0.7;
      });
      platforms.forEach((platform) => {
        platform.position.x += player.speed;
      });
    }
  }
  //this is the platform collision detection code
  platforms.forEach((platform) => {
    if (
      player.position.y + player.height <= platform.position.y &&
      player.position.y + player.height + player.velocity.y >=
        platform.position.y &&
      player.position.x + player.width >= platform.position.x &&
      player.position.x <= platform.position.x + platform.width
    ) {
      player.velocity.y = 0;
    }
  });

  if (player.position.y < 0) {
    player.velocity.y = 10;
  }

  if (
    keys.right.pressed &&
    lastKey === "right" &&
    player.currentSprite !== player.sprites.run.right
  ) {
    player.frames = 1;
    player.currentSprite = player.sprites.run.right;
  } else if (
    keys.left.pressed &&
    lastKey === "left" &&
    player.currentSprite !== player.sprites.run.left
  ) {
    player.frames = 1;
    player.currentSprite = player.sprites.run.left;
  } else if (
    !keys.left.pressed &&
    lastKey === "left" &&
    player.currentSprite !== player.sprites.idle.left
  ) {
    player.frames = 1;
    player.currentSprite = player.sprites.idle.left;
  } else if (
    !keys.right.pressed &&
    lastKey === "right" &&
    player.currentSprite !== player.sprites.idle.right
  ) {
    player.frames = 1;
    player.currentSprite = player.sprites.idle.right;
  }

  //win condition
  if (scrollOffset + canvas.width > 8300) {
    console.log("You Win!");

    youWinPhoto.draw();
    // ctx.font = "48px serif"
    // ctx.fillText("You Win!", 8300, 200, 400 )
  }
  //lose condition
  if (player.position.y > canvas.height) {
    console.log("you lose");
    start();
  }
}
start();
animate();

document.addEventListener("keydown", ({ keyCode }) => {
  //   console.log(event);
  switch (keyCode) {
    case 65:
      keys.left.pressed = true;
      lastKey = "left";
      break;
    case 37:
      keys.left.pressed = true;
      lastKey = "left";
      break;

    case 83:
      //   console.log("down");
      break;
    case 40:
      //   console.log("down");
      break;

    case 68:
      keys.right.pressed = true;
      lastKey = "right";
      break;
    case 39:
      keys.right.pressed = true;
      lastKey = "right";
      break;

    case 87:
      player.velocity.y -= 20;
      break;
    case 38:
      player.velocity.y -= 20;
      break;
  }
});

document.addEventListener("keyup", ({ keyCode }) => {
  //   console.log(event);
  switch (keyCode) {
    case 65:
      keys.left.pressed = false;
      break;
    case 37:
      keys.left.pressed = false;
      break;

    case 83:
      player.velocity.y = 0;
      break;
    case 40:
      player.velocity.y = 0;
      break;

    case 68:
      keys.right.pressed = false;
      break;
    case 39:
      keys.right.pressed = false;
      break;

    case 87:
      player.velocity.y = 0;
      break;
    case 38:
      player.velocity.y = 0;
      break;
  }
});
